contributors/design-proposals/device-plugin.md

```diff
@@ -0,0 +1,503 @@
+
+It is not a sustainable solution to expect every hardware vendor to add their
+vendor specific code inside Kubernetes to make their devices usable.
+
+Instead, we want a solution for vendors to be able to advertise their resources
+to Kubelet and monitor them without writing custom Kubernetes code.
+We also want to provide a consistent and portable solution for users to
+consume hardware devices across k8s clusters.
+
+This document describes a vendor independant solution to:
+  * Discovering and representing external devices
+  * Making these devices available to the containers, using these devices,
+    scrubbing and securly sharing these devices.
+  * Health Check of these devices
+
+Because devices are vendor dependant and have their own sets of problems
+and mechanisms, the solution we describe is a plugin mechanism that may run
+in a container deployed through the DaemonSets mechanism.
+The targeted devices include GPUs, High-performance NICs, FPGAs, InfiniBand,
+Storage devices, and other similar computing resources that require vendor
+specific initialization and setup.
+
+The goal is for a user to be able to enable vendor devices (e.g: GPUs) through
+the following simple steps:
+  * `kubectl create -f http://vendor.com/device-plugin-daemonset.yaml`
+  * When launching `kubectl describe nodes`, the devices appear in the node
+    status as `extensions.kubernetes.io/vendor-device`. Note: naming
+    convention is discussed PR [#844](https://github.com/kubernetes/community/pull/844)
+
+# Use Cases
+
+ * I want to use a particular device type (GPU, InfiniBand, FPGA, etc.)
+   in my pod.
+ * I should be able to use that device without writing custom Kubernetes code.
+ * I want a consistent and portable solution to consume hardware devices
+   across k8s clusters.
+
+# Objectives
+
+1. Add support for vendor specific Devices in kubelet:
+    * Through an extension mechanism.
+    * Which allows discovery and health check of devices.
+    * Which allows hooking the runtime to make devices available in containers
+      and cleaning them up.
+2. Define a deployment mechanism for this new API.
+3. Define a versioning mechanism for this new API.
+
+# Non Objectives
+
+1. Handling heterogeneous nodes and topology related problems
+2. Collecting metrics is not part of this proposal. We will only solve
+   Health Check.
+
+# TLDR
+
+At their core, device plugins are simple gRPC servers that may run in a
+container deployed through the pod mechanism.
+
+These servers implement the gRPC interface defined later in this design
+document and once the device plugin makes itself known to kubelet, kubelet
+will interact with the device through two simple functions:
+  1. A `ListAndWatch` function for the kubelet to Discover the devices and
+     their properties as well as notify of any status change (device
+     became unhealthy).
+  2. An `Allocate` function which is called before creating a user container
+     consuming any exported devices
+
```

Comment:
Once the Device is allocated and once container is created, there is a need to 'Bind' a newly allocated device to container before starting a container.
Such device in Linux could be character, block, network or combination of them. Infiniband device will be a combination of character & network device.
This requires access to mount namespace of container and host namespace to map a character device files and network namespace of host network namespace of container, so that device plugin can perform operations in that namespace.

So to summarize,
1. We need need additional Bind(), Unbind() calls.
2. We need pointers to possibly all namespaces of host and container to keep it generic.
3. Good to have cpu map that is assigned to a container, so that any cpu specific memory allocation optimizations can be applied.